src/graphs/TopologicalSort.re

```diff
@@ -0,0 +1,73 @@
+type node = {
+    id: string,
+    children: list(string),
+};
+
+type directedGraph = list(node);
+
+exception Invalid_node_id(string);
+
+exception Graph_not_DAG;
+
+let parseAdjList = adj_list => {
+    let adj_tbl = Hashtbl.create(List.length(adj_list));
+    let insert = ({id, children}) => {
+        Hashtbl.add(adj_tbl, id, children);
+    };
+
+    List.iter(insert, adj_list);
+    let validateChildren = node => {
+        List.iter(child_id => {
+            if (!Hashtbl.mem(adj_tbl, child_id)) {
+                raise(Invalid_node_id(child_id));
+            }
+        }, node.children);
+    };
+
+    List.iter(validateChildren, adj_list);
+    adj_tbl;
+};
+
+let rec visit = (~node_id, ~adj_tbl, ~visited, ~ordering, ~ancestors) => {
+    if (Hashtbl.mem(visited, node_id)) {
+        let pred = id => id == node_id;
+        switch (List.find(pred, ancestors)) {
+        | exception Not_found => ordering;
+        | _ => raise(Graph_not_DAG);
+        };
+    }
+    else {
+        let ancestors = List.append(ancestors, [node_id]);
+        let children = Hashtbl.find(adj_tbl, node_id);
+        let visitChild = (ordering, child_id) => {
+            visit(
```

Comment:
I don't like this pattern here. Let's create a record type for the arguments to this function `state`. We can then create a new state by replacing only parts of the state like

```
state with node_id = child_id
```